File: FaceBlurPixelate/yunet_webcam_face_blur.cpp
/*
YuNet Webcam Face Anonymizer (C++ Tutorial Version)

This example demonstrates a full real-time pipeline with OpenCV:
1) Open webcam stream.
2) Load YuNet neural face detector (FaceDetectorYN) from an ONNX model.
3) Detect faces on each frame.
4) Apply one of two anonymization methods on each detected face:
   - Gaussian blur
   - Pixelation / mosaic effect
5) Show the processed video feed live.

Why this file is heavily commented:
- The goal is to be instructional, so each major section explains not only
  "what" it does, but also "why" it is done this way.

Build example:
  g++ -std=c++17 yunet_webcam_face_blur.cpp -o yunet_webcam_face_blur `pkg-config --cflags --libs opencv4`

Run example:
  ./yunet_webcam_face_blur --model face_detection_yunet_2023mar.onnx --mode pixelate
*/

#include <filesystem>
#include <iostream>
#include <string>
#include <vector>

#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/objdetect/face.hpp>
#include <opencv2/videoio.hpp>

// Alias to reduce verbosity when using std::filesystem APIs.
namespace fs = std::filesystem;

/*
Configuration for the app.

This struct mirrors the CLI options so parameters are centralized and easy to
pass around.
*/
struct AppConfig {
    std::string model_path;       // Path to YuNet ONNX model; empty means auto-search.
    int camera_index = 0;         // Webcam index (0 is usually default camera).
    float score_threshold = 0.9f; // Minimum confidence for detections.
    float nms_threshold = 0.3f;   // NMS threshold to merge overlapping detections.
    int top_k = 5000;             // Max raw detections before NMS.
    bool show_score = false;      // If true, draw score text + rectangle.
    std::string mode = "blur";    // Anonymization mode: "blur" or "pixelate".
    int pixel_block_size = 16;    // Pixelation strength control.
};

// If --model is not provided, we try these filenames automatically.
static const std::vector<std::string> kDefaultModelCandidates = {
    "face_detection_yunet_2023mar.onnx",
    "face_detection_yunet_2022mar.onnx",
};

// Print command-line help.
static void printUsage(const char* prog) {
    std::cout
        << "YuNet Webcam Face Anonymizer (C++)\n\n"
        << "Usage:\n  " << prog
        << " [--model <path>] [--camera <idx>] [--score-threshold <float>]\n"
        << "      [--nms-threshold <float>] [--top-k <int>] [--show-score]\n"
        << "      [--mode blur|pixelate] [--pixel-block-size <int>]\n\n"
        << "Examples:\n"
        << "  " << prog << " --model face_detection_yunet_2023mar.onnx --mode blur\n"
        << "  " << prog << " --model face_detection_yunet_2023mar.onnx --mode pixelate --pixel-block-size 20\n";
}

/*
Find a value for args in the form:
  --key value

Returns true and writes out_value if key is found with a following token.
*/
static bool getArgValue(
    int argc,
    char** argv,
    const std::string& key,
    std::string& out_value
) {
    for (int i = 1; i < argc - 1; ++i) {
        if (argv[i] == key) {
            out_value = argv[i + 1];
            return true;
        }
    }
    return false;
}

// Return true if a boolean flag (e.g. --show-score) is present.
static bool hasFlag(int argc, char** argv, const std::string& flag) {
    for (int i = 1; i < argc; ++i) {
        if (argv[i] == flag) {
            return true;
        }
    }
    return false;
}

/*
Parse all CLI args into AppConfig.

Behavior notes:
- --help/-h prints usage and returns false (caller exits without error).
- Mode is validated to only allow blur/pixelate.
- Pixel block size is clamped to at least 1.
*/
static bool parseArgs(int argc, char** argv, AppConfig& cfg) {
    if (hasFlag(argc, argv, "--help") || hasFlag(argc, argv, "-h")) {
        printUsage(argv[0]);
        return false;
    }

    std::string value;

    // Optional model path.
    if (getArgValue(argc, argv, "--model", value)) {
        cfg.model_path = value;
    }

    // Camera index.
    if (getArgValue(argc, argv, "--camera", value)) {
        cfg.camera_index = std::stoi(value);
    }

    // Detector confidence threshold.
    if (getArgValue(argc, argv, "--score-threshold", value)) {
        cfg.score_threshold = std::stof(value);
    }

    // Detector NMS threshold.
    if (getArgValue(argc, argv, "--nms-threshold", value)) {
        cfg.nms_threshold = std::stof(value);
    }

    // Detector top-k.
    if (getArgValue(argc, argv, "--top-k", value)) {
        cfg.top_k = std::stoi(value);
    }

    // Processing mode.
    if (getArgValue(argc, argv, "--mode", value)) {
        cfg.mode = value;
    }

    // Pixelation strength.
    if (getArgValue(argc, argv, "--pixel-block-size", value)) {
        cfg.pixel_block_size = std::stoi(value);
    }

    // Bool flag.
    cfg.show_score = hasFlag(argc, argv, "--show-score");

    // Validate mode value explicitly.
    if (cfg.mode != "blur" && cfg.mode != "pixelate") {
        std::cerr << "Invalid --mode value: " << cfg.mode
                  << " (expected blur or pixelate)\n";
        return false;
    }

    // Keep this safe to avoid divide-by-zero in pixelation logic.
    if (cfg.pixel_block_size < 1) {
        cfg.pixel_block_size = 1;
    }

    return true;
}

/*
Resolve model path.

Search order:
1) If user passed --model, validate and use it.
2) Current working directory + default YuNet names.
3) Executable directory + default YuNet names.

Throws std::runtime_error if model cannot be found.
*/
static std::string resolveModelPath(const AppConfig& cfg, const char* argv0) {
    // Highest priority: explicit user path.
    if (!cfg.model_path.empty()) {
        fs::path p(cfg.model_path);
        p = fs::absolute(p);
        if (fs::is_regular_file(p)) {
            return p.string();
        }
        throw std::runtime_error("Model not found: " + p.string());
    }

    // Try current working directory.
    fs::path cwd = fs::current_path();
    for (const auto& name : kDefaultModelCandidates) {
        fs::path candidate = cwd / name;
        if (fs::is_regular_file(candidate)) {
            return candidate.string();
        }
    }

    // Try executable directory.
    fs::path exe_dir = fs::absolute(fs::path(argv0)).parent_path();
    for (const auto& name : kDefaultModelCandidates) {
        fs::path candidate = exe_dir / name;
        if (fs::is_regular_file(candidate)) {
            return candidate.string();
        }
    }

    throw std::runtime_error(
        "YuNet model not found. Pass --model <path> or place "
        "face_detection_yunet_2023mar.onnx / face_detection_yunet_2022mar.onnx "
        "in current directory."
    );
}

/*
Clamp a rectangle to frame bounds.

Face boxes can sometimes be partially outside the frame,
especially near image borders. This function protects ROI slicing.
*/
static bool clampRect(const cv::Rect& in, int frame_w, int frame_h, cv::Rect& out) {
    int x1 = std::max(0, in.x);
    int y1 = std::max(0, in.y);
    int x2 = std::min(frame_w, in.x + in.width);
    int y2 = std::min(frame_h, in.y + in.height);

    if (x2 <= x1 || y2 <= y1) {
        return false;
    }

    out = cv::Rect(x1, y1, x2 - x1, y2 - y1);
    return true;
}

/*
Apply Gaussian blur to a face region.

Kernel size is adapted based on face size so the effect is reasonably consistent
whether the face is small (far) or large (near).
*/
static void blurFaceRegion(cv::Mat& frame, const cv::Rect& rect) {
    cv::Mat roi = frame(rect);
    if (roi.empty()) {
        return;
    }

    int min_dim = std::min(rect.width, rect.height);

    // Base kernel from face size.
    int kernel = std::max(15, min_dim / 3);

    // Gaussian kernels must be odd values.
    if (kernel % 2 == 0) {
        kernel += 1;
    }

    cv::GaussianBlur(roi, roi, cv::Size(kernel, kernel), 0);
}

/*
Apply pixelation (mosaic) to a face region.

Classic trick:
1) Downscale aggressively -> detail removed.
2) Upscale with INTER_NEAREST -> visible square blocks.
*/
static void pixelateFaceRegion(cv::Mat& frame, const cv::Rect& rect, int block_size) {
    cv::Mat roi = frame(rect);
    if (roi.empty()) {
        return;
    }

    // Safety clamp.
    int safe_block = std::max(1, block_size);

    // Compute temporary low-res size.
    // Bigger block size => smaller temp image => stronger pixel effect.
    int small_w = std::max(1, roi.cols / safe_block);
    int small_h = std::max(1, roi.rows / safe_block);

    cv::Mat temp;

    // Shrink (loses detail).
    cv::resize(roi, temp, cv::Size(small_w, small_h), 0.0, 0.0, cv::INTER_LINEAR);

    // Expand with nearest-neighbor (keeps hard square blocks).
    cv::resize(temp, roi, roi.size(), 0.0, 0.0, cv::INTER_NEAREST);
}

int main(int argc, char** argv) {
    try {
        AppConfig cfg;

        // Parse CLI and potentially print help.
        if (!parseArgs(argc, argv, cfg)) {
            return 0;
        }

        // Locate ONNX model file.
        const std::string model_path = resolveModelPath(cfg, argv[0]);

        // Open webcam stream.
        cv::VideoCapture cap(cfg.camera_index);
        if (!cap.isOpened()) {
            std::cerr << "Cannot open webcam index " << cfg.camera_index << "\n";
            return 1;
        }

        // Read one frame first so we know actual resolution.
        // FaceDetectorYN requires an explicit input size at creation.
        cv::Mat frame;
        if (!cap.read(frame) || frame.empty()) {
            std::cerr << "Cannot read initial frame from webcam\n";
            return 1;
        }

        // Create YuNet detector instance.
        // API:
        // FaceDetectorYN::create(model, config, input_size, score_thresh, nms_thresh, top_k)
        cv::Ptr<cv::FaceDetectorYN> detector = cv::FaceDetectorYN::create(
            model_path,
            "",
            frame.size(),
            cfg.score_threshold,
            cfg.nms_threshold,
            cfg.top_k
        );

        std::cout << "Using model: " << model_path << "\n";
        std::cout << "Mode: " << cfg.mode << "\n";
        std::cout << "Press 'q' or ESC to quit\n";

        while (true) {
            // Grab frame.
            if (!cap.read(frame) || frame.empty()) {
                break;
            }

            // Keep detector input size synced with frame size.
            // Important if capture resolution changes.
            detector->setInputSize(frame.size());

            // Run detector.
            // Output matrix format: N x 15 float values per detected face:
            // [x, y, w, h, l0x, l0y, l1x, l1y, l2x, l2y, l3x, l3y, l4x, l4y, score]
            cv::Mat faces;
            detector->detect(frame, faces);

            // Iterate over all detections.
            for (int i = 0; i < faces.rows; ++i) {
                const cv::Mat row = faces.row(i);

                int x = static_cast<int>(row.at<float>(0));
                int y = static_cast<int>(row.at<float>(1));
                int w = static_cast<int>(row.at<float>(2));
                int h = static_cast<int>(row.at<float>(3));
                float score = row.at<float>(14);

                // Ensure rectangle is valid and inside frame bounds.
                cv::Rect clamped;
                if (!clampRect(cv::Rect(x, y, w, h), frame.cols, frame.rows, clamped)) {
                    continue;
                }

                // Apply selected anonymization mode.
                if (cfg.mode == "pixelate") {
                    pixelateFaceRegion(frame, clamped, cfg.pixel_block_size);
                } else {
                    blurFaceRegion(frame, clamped);
                }

                // Optional score display to help tune thresholds.
                if (cfg.show_score) {
                    cv::rectangle(frame, clamped, cv::Scalar(40, 255, 40), 2);
                    cv::putText(
                        frame,
                        cv::format("%.2f", score),
                        cv::Point(clamped.x, std::max(0, clamped.y - 8)),
                        cv::FONT_HERSHEY_SIMPLEX,
                        0.6,
                        cv::Scalar(40, 255, 40),
                        2,
                        cv::LINE_AA
                    );
                }
            }

            // Display processed frame.
            cv::imshow("YuNet Face Blur/Pixelate (C++)", frame);

            // waitKey(1) also handles UI event pumping for HighGUI windows.
            int key = cv::waitKey(1) & 0xFF;
            if (key == 27 || key == 'q') {
                break;
            }
        }

        // Cleanup.
        cap.release();
        cv::destroyAllWindows();
        return 0;
    } catch (const std::exception& e) {
        // Centralized error handling for argument/model/path issues.
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: FaceBlurPixelate/yunet_webcam_face_blur.py
"""
YuNet Webcam Face Anonymizer (Tutorial Version)

This script shows how to:
1) Load OpenCV's YuNet neural face detector (FaceDetectorYN).
2) Read frames from a webcam in real time.
3) Detect faces in each frame.
4) Apply one of two anonymization effects on each detected face:
   - Gaussian blur
   - Pixelation (mosaic)
5) Display the processed stream live.

Typical run command:
    python yunet_webcam_face_blur.py \
        --model face_detection_yunet_2023mar.onnx \
        --mode pixelate
"""

import argparse
from pathlib import Path

import cv2


# If the user does not pass --model, we try these filenames in the script folder.
# Both are official YuNet model filenames commonly seen in tutorials/repos.
DEFAULT_MODEL_CANDIDATES = [
    "face_detection_yunet_2023mar.onnx",
    "face_detection_yunet_2022mar.onnx",
]


def parse_args() -> argparse.Namespace:
    """Define and parse command-line arguments.

    Keeping parameters configurable from CLI makes the tutorial practical:
    users can tune detector confidence, camera index, and anonymization mode
    without editing source code.
    """
    parser = argparse.ArgumentParser(
        description="Real-time webcam face detection + blur/pixelate using OpenCV YuNet"
    )

    # Path to ONNX model. If omitted, we auto-search common names.
    parser.add_argument(
        "--model",
        type=str,
        default=None,
        help="Path to YuNet ONNX model. If omitted, script tries common filenames in this folder.",
    )

    # Webcam index (0 is usually default webcam, 1/2... for external cameras).
    parser.add_argument("--camera", type=int, default=0, help="Webcam index (default: 0)")

    # Minimum confidence for detected face boxes.
    # Increase to reduce false positives; decrease if detections are missed.
    parser.add_argument(
        "--score-threshold",
        type=float,
        default=0.9,
        help="Minimum face confidence score (default: 0.9)",
    )

    # NMS threshold (Non-Maximum Suppression) merges overlapping detections.
    parser.add_argument(
        "--nms-threshold",
        type=float,
        default=0.3,
        help="NMS threshold for YuNet (default: 0.3)",
    )

    # Max number of raw detections before NMS.
    parser.add_argument(
        "--top-k",
        type=int,
        default=5000,
        help="Maximum number of detections before NMS (default: 5000)",
    )

    # Optional visualization aid: draw confidence scores.
    parser.add_argument(
        "--show-score",
        action="store_true",
        help="Draw detection score on each face",
    )

    # Choose anonymization method.
    parser.add_argument(
        "--mode",
        type=str,
        choices=["blur", "pixelate"],
        default="blur",
        help="Face anonymization mode: blur or pixelate (default: blur)",
    )

    # Pixelation control: bigger block => stronger mosaic effect.
    parser.add_argument(
        "--pixel-block-size",
        type=int,
        default=16,
        help="Pixel block size for pixelation mode (default: 16)",
    )

    return parser.parse_args()


def resolve_model_path(cli_path: str | None) -> Path:
    """Resolve and validate the YuNet model path.

    Priority:
    1) If user passed --model, use it.
    2) Otherwise look for known filenames in the script directory.
    """
    if cli_path:
        model_path = Path(cli_path).expanduser().resolve()
        if model_path.is_file():
            return model_path
        raise FileNotFoundError(f"Model not found: {model_path}")

    script_dir = Path(__file__).resolve().parent
    for name in DEFAULT_MODEL_CANDIDATES:
        candidate = script_dir / name
        if candidate.is_file():
            return candidate

    expected = " or ".join(DEFAULT_MODEL_CANDIDATES)
    raise FileNotFoundError(
        "YuNet model not found. Pass --model <path-to-onnx> or place "
        f"{expected} in {script_dir}"
    )


def clamp_rect(x: int, y: int, w: int, h: int, frame_w: int, frame_h: int):
    """Clip a rectangle to frame boundaries.

    Detection boxes can occasionally touch/overflow edges.
    Clamping prevents slicing errors when extracting face ROI.
    Returns None if resulting box is invalid (zero or negative area).
    """
    x1 = max(0, x)
    y1 = max(0, y)
    x2 = min(frame_w, x + w)
    y2 = min(frame_h, y + h)

    if x2 <= x1 or y2 <= y1:
        return None

    return x1, y1, x2, y2


def blur_face_region(frame, rect):
    """Apply Gaussian blur to one detected face ROI.

    We scale kernel size with face size so blur intensity looks consistent
    across near/far faces.
    """
    x1, y1, x2, y2 = rect

    # Region of interest (face crop from the frame).
    roi = frame[y1:y2, x1:x2]
    if roi.size == 0:
        return

    face_w = x2 - x1
    face_h = y2 - y1

    # Kernel size for Gaussian blur must be odd.
    # We derive it from face dimensions for adaptive strength.
    kernel = max(15, ((min(face_w, face_h) // 3) | 1))

    blurred = cv2.GaussianBlur(roi, (kernel, kernel), 0)
    frame[y1:y2, x1:x2] = blurred


def pixelate_face_region(frame, rect, block_size: int):
    """Apply mosaic pixelation to one detected face ROI.

    Technique:
    1) Downscale ROI heavily.
    2) Upscale back with nearest-neighbor interpolation.
    This creates large square color blocks.
    """
    x1, y1, x2, y2 = rect
    roi = frame[y1:y2, x1:x2]
    if roi.size == 0:
        return

    h, w = roi.shape[:2]

    # Guard against invalid block size and tiny ROIs.
    safe_block = max(1, block_size)

    # Small intermediate image dimensions.
    # Larger block size -> smaller intermediate image -> stronger pixelation.
    small_w = max(1, w // safe_block)
    small_h = max(1, h // safe_block)

    # First resize: compress details.
    temp = cv2.resize(roi, (small_w, small_h), interpolation=cv2.INTER_LINEAR)

    # Second resize: expand with nearest neighbor to preserve "blocks".
    pixelated = cv2.resize(temp, (w, h), interpolation=cv2.INTER_NEAREST)

    # Write processed ROI back into original frame.
    frame[y1:y2, x1:x2] = pixelated


def main() -> None:
    """Main application loop."""
    args = parse_args()
    model_path = resolve_model_path(args.model)

    # Open camera stream.
    cap = cv2.VideoCapture(args.camera)
    if not cap.isOpened():
        raise RuntimeError(f"Cannot open webcam index {args.camera}")

    # Read one frame first so we know actual camera resolution.
    # YuNet detector needs input size configured explicitly.
    ok, frame = cap.read()
    if not ok or frame is None:
        cap.release()
        raise RuntimeError("Cannot read initial frame from webcam")

    frame_h, frame_w = frame.shape[:2]

    # Create YuNet detector.
    # Signature: FaceDetectorYN.create(model, config, input_size, score_thr, nms_thr, top_k)
    detector = cv2.FaceDetectorYN.create(
        str(model_path),
        "",
        (frame_w, frame_h),
        args.score_threshold,
        args.nms_threshold,
        args.top_k,
    )

    print(f"Using model: {model_path}")
    print(f"Mode: {args.mode}")
    print("Press 'q' or ESC to quit")

    while True:
        # Read next webcam frame.
        ok, frame = cap.read()
        if not ok or frame is None:
            # End loop gracefully if camera stops providing frames.
            break

        h, w = frame.shape[:2]

        # Important: keep detector input size in sync with frame size.
        # If resolution changes, detector must be updated.
        detector.setInputSize((w, h))

        # Run neural face detection.
        # faces is either None or an ndarray shaped (N, 15):
        # [x, y, w, h, 5 landmarks (10 values), score]
        _, faces = detector.detect(frame)

        if faces is not None:
            for face in faces:
                # Face bounding box is first 4 values.
                x, y, bw, bh = face[:4].astype(int)

                # Keep box inside frame to avoid slicing issues.
                rect = clamp_rect(x, y, bw, bh, w, h)
                if rect is None:
                    continue

                # Apply selected anonymization effect.
                if args.mode == "pixelate":
                    pixelate_face_region(frame, rect, args.pixel_block_size)
                else:
                    blur_face_region(frame, rect)

                # Optional score overlay for debugging/tuning thresholds.
                if args.show_score:
                    score = float(face[-1])
                    x1, y1, x2, y2 = rect
                    cv2.rectangle(frame, (x1, y1), (x2, y2), (40, 255, 40), 2)
                    cv2.putText(
                        frame,
                        f"{score:.2f}",
                        (x1, max(0, y1 - 8)),
                        cv2.FONT_HERSHEY_SIMPLEX,
                        0.6,
                        (40, 255, 40),
                        2,
                        cv2.LINE_AA,
                    )

        # Show processed frame in a window.
        cv2.imshow("YuNet Face Blur/Pixelate", frame)

        # waitKey(1) processes UI events and captures keyboard input.
        key = cv2.waitKey(1) & 0xFF
        if key in (27, ord("q")):
            break

    # Always release resources cleanly.
    cap.release()
    cv2.destroyAllWindows()


if __name__ == "__main__":
    main()
